language/move-stdlib/src/natives/debug.rs

```diff
@@ -3,20 +3,537 @@
 // SPDX-License-Identifier: Apache-2.0
 
 use crate::natives::helpers::make_module_natives;
+#[cfg(feature = "testing")]
+use move_binary_format::errors::PartialVMError;
 use move_binary_format::errors::PartialVMResult;
+use move_core_types::{account_address::AccountAddress, gas_algebra::InternalGas};
+#[cfg(feature = "testing")]
 use move_core_types::{
-    account_address::AccountAddress, gas_algebra::InternalGas, language_storage::TypeTag,
+    language_storage::{StructTag, TypeTag},
+    value::{MoveStruct, MoveStructLayout, MoveTypeLayout, MoveValue},
+    vm_status::StatusCode,
 };
 use move_vm_runtime::native_functions::{NativeContext, NativeFunction};
-use move_vm_types::values::VectorRef;
-#[allow(unused_imports)]
-use move_vm_types::values::{values_impl::debug::print_reference, Reference, StructRef};
+#[cfg(feature = "testing")]
+use move_vm_types::values::Reference;
 #[allow(unused_imports)]
 use move_vm_types::{
     loaded_data::runtime_types::Type, natives::function::NativeResult, pop_arg, values::Value,
 };
 use smallvec::smallvec;
 use std::{collections::VecDeque, sync::Arc};
+#[cfg(feature = "testing")]
+use std::{
+    fmt,
+    fmt::{Error, Write},
+};
+
+#[cfg(feature = "testing")]
+fn is_string_struct_tag(struct_tag: &StructTag, move_std_addr: &AccountAddress) -> bool {
+    struct_tag.address == *move_std_addr
+        && struct_tag.module.as_str().eq("string")
+        && struct_tag.name.as_str().eq("String")
+}
+
+#[cfg(feature = "testing")]
+const HANDLE_FMT_WRITE_ERROR: fn(Error) -> PartialVMError = |e: fmt::Error| {
+    let vmerr = PartialVMError::new(StatusCode::UNKNOWN_STATUS);
+    vmerr.with_message("write! macro failed with: ".to_string() + e.to_string().as_str())
+};
+
+#[cfg(feature = "testing")]
+fn get_annotated_struct_layout(
+    context: &NativeContext,
+    ty: &Type,
+) -> PartialVMResult<MoveStructLayout> {
+    let annotated_type_layout = context.type_to_fully_annotated_layout(ty)?.unwrap();
+    match annotated_type_layout {
+        MoveTypeLayout::Struct(annotated_struct_layout) => Ok(annotated_struct_layout),
+        _ => Err(
+            PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR).with_message(
+                "Could not convert Type to fully-annotated MoveTypeLayout via NativeContext"
+                    .to_string(),
+            ),
+        ),
+    }
+}
+#[cfg(feature = "testing")]
+fn get_vector_inner_type(ty: &Type) -> PartialVMResult<&Type> {
+    match ty {
+        Type::Vector(ty) => Ok(ty),
+        _ => Err(PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR)
+            .with_message("Could not get the inner Type of a vector's Type".to_string())),
+    }
+}
+
+#[cfg(feature = "testing")]
+fn vector_move_value_to_vec_u8(bytes: Vec<MoveValue>) -> PartialVMResult<Vec<u8>> {
```

Comment:
Here too, might be useful in `impl MoveValue` (w/ or w/o the testing gate)

Reply:
Will move.